ext/http/00_serve.js

```diff
@@ -0,0 +1,466 @@
+const _wantsUpgrade = Symbol("_wantsUpgrade");
+const _upgraded = Symbol("_upgraded");
+
+class RequestHeaders {
+  #slabId;
+
+  get(header) {
+    const value = core.ops.op_get_request_header(this.#slabId, header);
+    return value;
+  }
+}
+
+class InnerRequest {
+  #slabId;
+  #headers;
+  #context;
+  #methodAndUri;
+  #streamRid;
+  #body;
+  #upgraded;
+
+  constructor(slabId, context) {
+    this.#slabId = slabId;
+    this.#context = context;
+    this.#upgraded = false;
+  }
+
+  close() {
+    if (this.#streamRid !== undefined) {
+      core.close(this.#streamRid);
+      this.#streamRid = undefined;
+    }
+    this.#slabId = undefined;
+  }
+
+  get [_upgraded]() {
+    return this.#upgraded;
+  }
+
+  [_wantsUpgrade](upgradeType, ...originalArgs) {
+    this.#upgraded = true;
+
+    // upgradeHttp is async
+    // TODO(mmastrac)
+    if (upgradeType == "upgradeHttp") {
+      throw "upgradeHttp is unavailable in Deno.serve at this time";
+    }
+
+    // upgradeHttpRaw is async
+    // TODO(mmastrac)
+    if (upgradeType == "upgradeHttpRaw") {
+      throw "upgradeHttp is unavailable in Deno.serve at this time";
+    }
+
+    // upgradeWebSocket is sync
+    if (upgradeType == "upgradeWebSocket") {
+      const response = originalArgs[0];
+      const ws = originalArgs[1];
+
+      this.url();
+      this.headerList;
+      this.close();
+
+      // Start the upgrade in the background.
+      (async () => {
+        try {
+          const upgrade = await core.opAsync2(
+            "op_upgrade",
+            this.#slabId,
+            response.headerList,
+          );
+          const wsRid = core.ops.op_server_create(upgrade);
+          ws[_rid] = wsRid;
+          ws[_readyState] = WebSocket.OPEN;
+          ws[_role] = SERVER;
+          const event = new Event("open");
+          ws.dispatchEvent(event);
+
+          ws[_eventLoop]();
+          if (ws[_idleTimeoutDuration]) {
+            ws.addEventListener(
+              "close",
+              () => clearTimeout(ws[_idleTimeoutTimeout]),
+            );
+          }
+          ws[_serverHandleIdleTimeout]();
+        } catch (error) {
+          const event = new ErrorEvent("error", { error });
+          ws.dispatchEvent(event);
+        }
+      })();
+      return { response: new Response(), socket: ws };
+    }
+  }
+
+  url() {
+    if (this.#methodAndUri === undefined) {
+      if (this.#slabId === undefined) {
+        throw new TypeError("request closed");
+      }
+      this.#methodAndUri = core.ops.op_get_request_method_and_url(this.#slabId);
+    }
+    if (this.#methodAndUri[1]) {
+      return this.#context.scheme + this.#methodAndUri[1] +
+        this.#methodAndUri[2];
+    }
+    return this.#context.scheme + this.#methodAndUri[2];
+  }
+
+  get remoteAddr() {
+    if (this.#methodAndUri === undefined) {
+      if (this.#slabId === undefined) {
+        throw new TypeError("request closed");
+      }
```

Comment:
It is exceedingly likely that this will break existing code, because significant existing code using the existing HTTP API relies on the fact that `headers` and `url` etc can be accessed even after the response is sent. An example is the GA4 handler in dotland.

Has there been any investigative work on this?